tests/integration2/tunnel/tunnel_test.go

```diff
@@ -0,0 +1,344 @@
+			"ca.crt":      readFileOrFail("certs/ca.crt", t),
+			"service.crt": readFileOrFail("certs/service.crt", t),
+			"service.key": readFileOrFail("certs/service.key", t),
+			"client.crt":  readFileOrFail("certs/client.crt", t),
+			"client.key":  readFileOrFail("certs/client.key", t),
+		}}
+	_ = egress.NewOrFail(t, ctx, egress.Config{
+		Istio:                      ist,
+		Secret:                     secret,
+		AdditionalSecretMountPoint: "/etc/istio/tunnel-certs",
+	})
+
+	ingress := ingress.NewOrFail(t, ctx, ingress.Config{
+		Istio:                      ist,
+		Secret:                     secret,
+		AdditionalSecretMountPoint: "/etc/istio/tunnel-certs",
+	})
+
+	pilot := pilot.NewOrFail(t, ctx, pilot.Config{})
+	applications := apps.NewOrFail(ctx, t, apps.Config{Pilot: pilot})
+	a := applications.GetAppOrFail("a", t)
+	b := applications.GetAppOrFail("t", t).(*apps.KubeApp)
+
+	be := b.EndpointsForProtocol(model.ProtocolHTTP)[0].(*apps.KubeEndpoint)
+
+	ingressURL, err := ingress.URL(model.ProtocolHTTPS)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ingressPort := ingressURL.Port()
+
+	beURL := be.URL()
+	virtualPort := 8080
+	serviceName := "client"
+	env := ctx.Environment().(*kube.Environment)
+	cfg, err := istio.DefaultConfig(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	clientNamespace := namespace.NewOrFail(t, ctx, "client", true)
+	virtualIP := env.AllocateIPAddressOrFail(virtualPort, serviceName, clientNamespace.Name(), t)
+	serverNamespace := namespace.NewOrFail(t, ctx, "server", true)
+
+	err = env.ApplyContents(cfg.SystemNamespace,
+		dump(tmpl.EvaluateOrFail(t, clientSideEgressConfig, map[string]interface{}{
+			"ingressAddress":  ingressURL.Hostname(),
+			"ingressPort":     ingressPort,
+			"ingressDNS":      "service.istio.test.local", // Must match CN in certs/server.crt
+			"sidecarSNI":      "sni.of.destination.rule.in.sidecar",
+			"systemNamespace": cfg.SystemNamespace,
+		})))
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = env.ApplyContents(clientNamespace.Name(),
+		dump(tmpl.EvaluateOrFail(t, clientSideConfig, map[string]interface{}{
+			"vip":             virtualIP,
+			"serviceName":     serviceName,
+			"sidecarSNI":      "sni.of.destination.rule.in.sidecar",
+			"systemNamespace": cfg.SystemNamespace,
+		})),
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = env.ApplyContents(serverNamespace.Name(),
+		dump(tmpl.EvaluateOrFail(t, serverSideConfig, map[string]interface{}{
+			"address":    b.ClusterIP(),
+			"port":       beURL.Port(),
+			"ingressDNS": "service.istio.test.local", // Must match CN in certs/server.crt
+			"clientSAN":  "client.istio.test.local",  // Must match CN and SAN in certs/client.crt
+		})),
+	)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	log.Infof("wait for 10 seconds for config distribution.") // see https://github.com/istio/istio/issues/6170
+	time.Sleep(10 * time.Second)
+
+	be.SetURL(&url.URL{Host: fmt.Sprintf("%s:%d", virtualIP, virtualPort), Path: beURL.Path, Scheme: beURL.Scheme})
```

Comment:
Can you put this in a retry loop, instead of sleeping? That will ensure that the test will not be affected by timing issues.

Reply:
OK